crates/league-modpkg/src/metadata.rs

```diff
@@ -0,0 +1,151 @@
+use binrw::binrw;
+
+use crate::{
+    license::ModpkgLicense,
+    utils::{optional_string_len, optional_string_read, optional_string_write},
+};
+use serde::{Deserialize, Serialize};
+
+#[binrw]
+#[brw(little)]
+#[derive(Debug, PartialEq, Default)]
+#[cfg_attr(test, derive(proptest_derive::Arbitrary))]
+pub struct ModpkgMetadata {
+    #[br(temp)]
+    #[bw(calc = name.len() as u32)]
+    name_len: u32,
+    #[br(count = name_len, try_map = String::from_utf8)]
+    #[bw(map = |s| s.as_bytes().to_vec())]
+    name: String,
+
+    #[br(temp)]
+    #[bw(calc = display_name.len() as u32)]
+    display_name_len: u32,
+    #[br(count = display_name_len, try_map = String::from_utf8)]
+    #[bw(map = |s| s.as_bytes().to_vec())]
+    display_name: String,
+
+    #[br(temp)]
+    #[bw(calc = optional_string_len(description) as u32)]
+    description_len: u32,
+    #[brw(if(description_len > 0))]
+    #[bw(map = optional_string_write)]
+    #[br(count = description_len, try_map = optional_string_read)]
+    description: Option<String>,
+
+    #[br(temp)]
+    #[bw(calc = version.len() as u32)]
+    version_len: u32,
+    #[br(count = version_len, try_map = String::from_utf8)]
+    #[bw(map = |s| s.as_bytes().to_vec())]
+    version: String,
+
+    #[br(temp)]
+    #[bw(calc = optional_string_len(distributor) as u32)]
+    distributor_len: u32,
+    #[brw(if(distributor_len> 0))]
+    #[bw(map = optional_string_write)]
+    #[br(count = distributor_len, try_map = optional_string_read)]
+    distributor: Option<String>,
+
+    #[br(temp)]
+    #[bw(calc = (authors.len()) as u32)]
+    author_count: u32,
+    #[br(count = author_count)]
+    authors: Vec<ModpkgAuthor>,
+
+    license: ModpkgLicense,
+}
+
+impl ModpkgMetadata {
+    /// The total size of the metadata when written to bytes.
+    pub fn size(&self) -> usize {
```

Comment:
There should be better way to do this with binrw

Reply:
As far as I know, we either do this or we use some custom String struct that impls BinWrite/Read (or we ask binrw to implement mapping inside an option somehow)

Reply:
Why not just implement dummy Write that tracks the size and write metdata into it?

Reply:
Github mobile lied to me, thought this was a comment on the optional_string_write map stuff. There probably is a better way I just ported what already was there - I'll let @Crauzer decide what to replace it with

Reply:
Honestly, I don't even see the point of this function existing now that we're using `binrw`. I think we can remove it. If needed, we can add it back using some other way as moon said.